### PR TITLE
add code to push an image

### PR DESCRIPTION
**Pull request by abdelghanimeliani**
…de-in-your-tool tutorial

#### What type of PR is this?

/kind other

#### How to verify it
just test the code

### CODE REVIEW

File: docs/tutorials/04-include-in-your-build-tool.md [hunk 1]
```diff
@@ -95,6 +95,40 @@ Now you can run commit the build:
 imageId, _, _, err := builder.Commit(context.TODO(), imageRef, buildah.CommitOptions{})
 ```
 
+## Push the image
+first, you have to add another image reference from the `alltransports` package to refer the destination image registry.
```
Comment (rhatdan) [inline docs/tutorials/04-include-in-your-build-tool.md:99]: remove "you have to"

File: docs/tutorials/04-include-in-your-build-tool.md [hunk 2]
```diff
@@ -95,6 +95,40 @@ Now you can run commit the build:
...
+Next , you have to create a `types.DockerAuthConfig` object to authenticate your tool against the registry.
```
Comment (rhatdan) [inline docs/tutorials/04-include-in-your-build-tool.md:100]: remove "you have to"

**Comment by rhatdan** [comment]
Please sign you commits

git commit -a --amend -s
git push --force

**Comment by abdelghanimeliani** [comment]
> Please sign you commits
> 
> git commit -a --amend -s git push --force

@rhatdan done

**Comment by rhatdan** [comment]
@vrothberg @mtrmac PTAL

File: docs/tutorials/04-include-in-your-build-tool.md [hunk 3]
```diff
@@ -95,6 +95,40 @@ Now you can run commit the build:
 imageId, _, _, err := builder.Commit(context.TODO(), imageRef, buildah.CommitOptions{})
 ```
 
+## Push the image
+first, add another image reference from the `alltransports` package to refer the destination image registry.
```
Comment (TomSweeneyRedHat) [inline docs/tutorials/04-include-in-your-build-tool.md:99]:
```suggestion
First, add another image reference from the `alltransports` package to refer to the destination image registry.
```

File: docs/tutorials/04-include-in-your-build-tool.md [hunk 4]
```diff
@@ -95,6 +95,40 @@ Now you can run commit the build:
...
+Next , create a `types.DockerAuthConfig` object to authenticate your tool against the registry.
```
Comment (TomSweeneyRedHat) [inline docs/tutorials/04-include-in-your-build-tool.md:100]:
```suggestion
Next, create a `types.DockerAuthConfig` object to authenticate your tool against the registry.
```

File: docs/tutorials/04-include-in-your-build-tool.md [hunk 5]
```diff
@@ -95,6 +95,40 @@ Now you can run commit the build:
...
+	if err != nil {
+
+		fmt.Println(err)
+	}
+
+	// Build an image scratch
+	s1, s2, err := buildah.Push(context.TODO(), "", destImageRef, buildah.PushOptions{
+		SystemContext: &types.SystemContext{
+			DockerAuthConfig: &types.DockerAuthConfig{
+				Username: "your password",
```
Comment (TomSweeneyRedHat) [inline docs/tutorials/04-include-in-your-build-tool.md:116]:
```suggestion
				Username: "your username",
```

File: docs/tutorials/04-include-in-your-build-tool.md [hunk 6]
```diff
@@ -199,5 +234,31 @@ func main() {
   }
 
   fmt.Printf("Image built! %s\n", imageId)
+
+  //if the registry isn't docker hub you should add docker:// prefix to the image name
```
Comment (TomSweeneyRedHat) [inline docs/tutorials/04-include-in-your-build-tool.md:238]:
```suggestion
  //if the registry isn't Docker Hub, you should add the docker:// prefix to the image name
```

File: docs/tutorials/04-include-in-your-build-tool.md [hunk 7]
```diff
@@ -199,5 +234,31 @@ func main() {
...
+	if err != nil {
+
+		fmt.Println(err)
+	}
+
+	// Build an image scratch
+	s1, s2, err := buildah.Push(context.TODO(), "", destImageRef, buildah.PushOptions{
+		SystemContext: &types.SystemContext{
+			DockerAuthConfig: &types.DockerAuthConfig{
+				Username: "your password",
```
Comment (TomSweeneyRedHat) [inline docs/tutorials/04-include-in-your-build-tool.md:250]:
```suggestion
				Username: "your username",
```

**Review by abdelghanimeliani — COMMENTED** [review]
@TomSweeneyRedHat done

**Comment by TomSweeneyRedHat** [comment]
@abdelghanimeliani I only see the first comment changed, and the others resolved without a change or comment?  i.e. "docker hub" to "Docker Hub" (we've been burned by that in the past),  and then username instead of password, unless I'm really misunderstanding the code sample.  Thx!

**Comment by abdelghanimeliani** [comment]
@TomSweeneyRedHat no it was my bad, it is the username

**Comment by abdelghanimeliani** [comment]
the user must add the password too , i will fix that and push again

File: docs/tutorials/04-include-in-your-build-tool.md [hunk 8]
```diff
@@ -95,6 +95,41 @@ Now you can run commit the build:
...
 ```
 
+## Push the image
+First, add another image reference from the `alltransports` package to refer to the destination image registry.
+Next , create a `types.DockerAuthConfig` object to authenticate your tool against the registry.
+
+```go
+
+//if the registry isn't Docker Hub, you should add the docker:// prefix to the image name
+destImageRef, err := alltransports.ParseImageName("docker.io/myusername/my-image")
```
Comment (mtrmac) [inline docs/tutorials/04-include-in-your-build-tool.md:105]: This doesn’t work at all! That makes me question whether the parts of the Buildah API (which I didn’t check) work.

---

If the goal is to push to registry, using `docker.ParseReference("//registry.example/…")` would be a bit simpler (and _potentially_ might eventually result in fewer dependencies)

**Review by mtrmac — COMMENTED** [review]
This is not really a review, more of a request for experts to take a more careful look.

Ideally, someone might want to test the code in practice…

File: docs/tutorials/04-include-in-your-build-tool.md (hunk 8 above)
Comment (abdelghanimeliani) [reply docs/tutorials/04-include-in-your-build-tool.md:105]: i don't know about docker.ParseReference("//registry.example/…"), i tried to push some images to quay image registry with this code and it works fine, can i see your output?

Comment (abdelghanimeliani) [reply docs/tutorials/04-include-in-your-build-tool.md:105]: And actually, this is the same code used when running **buildah push source destination** : https://github.com/containers/buildah/blob/main/cmd/buildah/push.go#L150

Comment (mtrmac) [reply docs/tutorials/04-include-in-your-build-tool.md:105]: > i tried to push some images to quay image registry with this code and it works fine

That’s, bluntly, impossible. `ParseImageName` fails on any input without a `:`.

---

See all the heuristics in https://github.com/containers/buildah/blob/63a7435736f685ca998d6823948c5714c78a71c6/cmd/buildah/push.go#L152 ; that’s what makes such input acceptable to `buildah push`.

File: docs/tutorials/04-include-in-your-build-tool.md [hunk 9]
```diff
@@ -199,5 +235,32 @@ func main() {
   }
 
   fmt.Printf("Image built! %s\n", imageId)
+
+  //if the registry isn't Docker Hub you should add the docker:// prefix to the image name
+destImageRef, err := alltransports.ParseImageName("docker.io/myusername/my-image")
```
[the remark this answers is not in the document]
Comment (abdelghanimeliani) [reply docs/tutorials/04-include-in-your-build-tool.md:240]: Clear, this is the same precedant code putted in one snippet with the others parts, it is the same idea.

File: docs/tutorials/04-include-in-your-build-tool.md (hunk 8 above)
Comment (abdelghanimeliani) [reply docs/tutorials/04-include-in-your-build-tool.md:105]: @mtrmac 
Honestly, no idea about this details but im using this code in kubernetes operator and it works fine
I am pushing an image with the name **"localhost/built_with_oprator"** from this destination **"quay.io/abdelghanimeliani/restore-counter:from-the-op-running-in-pod"**and this is the link : https://quay.io/repository/abdelghanimeliani/restore-counter?tab=tags


image

File: docs/tutorials/04-include-in-your-build-tool.md [hunk 10]
```diff
@@ -114,6 +114,7 @@ destImageRef, err := alltransports.ParseImageName("docker.io/myusername/my-image
 		SystemContext: &types.SystemContext{
 			DockerAuthConfig: &types.DockerAuthConfig{
 				Username: "your username",
+        Password: "your password",
```
Comment (TomSweeneyRedHat) [inline docs/tutorials/04-include-in-your-build-tool.md:117]: This is a possible nit of a nit, but there seems to be a space vs tab issue on this particular line.

**Comment by abdelghanimeliani** [comment]
@TomSweeneyRedHat , yes you are right , i fixed that

**Comment by rhatdan** [comment]
@abdelghanimeliani Are you still interested in this PR?

If so please fix the commit message to pass check.

@flouthoc PTAL

**Comment by abdelghanimeliani** [comment]
okey i will close this PR for now, if anyone is interested with the code in this branch.

**Pull request closed without merging**